File: app/src/main/java/me/zhanghai/android/files/provider/linux/LinuxFileStore.java
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.linux;

import android.os.Parcel;

import java.io.IOException;

import androidx.annotation.NonNull;
import java8.nio.file.Path;
import me.zhanghai.android.files.provider.root.RootPosixFileStore;
import me.zhanghai.android.files.provider.root.RootablePosixFileStore;

class LinuxFileStore extends RootablePosixFileStore {

    @NonNull
    private final LinuxPath mPath;
    @NonNull
    private final LocalLinuxFileStore mLocalFileStore;

    public LinuxFileStore(@NonNull LinuxPath path) throws IOException {
        this(path, new LocalLinuxFileStore(path));
    }

    private LinuxFileStore(@NonNull LinuxPath path, @NonNull LocalLinuxFileStore localFileStore) {
        super(path, localFileStore, RootPosixFileStore::new);

        mPath = path;
        mLocalFileStore = localFileStore;
    }


    public static final Creator<LinuxFileStore> CREATOR = new Creator<LinuxFileStore>() {
        @Override
        public LinuxFileStore createFromParcel(Parcel source) {
            return new LinuxFileStore(source);
        }
        @Override
        public LinuxFileStore[] newArray(int size) {
            return new LinuxFileStore[size];
        }
    };

    protected LinuxFileStore(Parcel in) {
        this(in.readParcelable(Path.class.getClassLoader()),
                in.readParcelable(LocalLinuxFileStore.class.getClassLoader()));
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(mPath, flags);
        dest.writeParcelable(mLocalFileStore, flags);
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/linux/LinuxFileStore.kt
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.linux

import android.os.Parcel
import android.os.Parcelable
import me.zhanghai.android.files.provider.root.RootPosixFileStore
import me.zhanghai.android.files.provider.root.RootablePosixFileStore

internal class LinuxFileStore private constructor(
    private val path: LinuxPath,
    private val localFileStore: LocalLinuxFileStore
) : RootablePosixFileStore(path, localFileStore, { RootPosixFileStore(it) }) {
    constructor(path: LinuxPath) : this(path, LocalLinuxFileStore(path))

    private constructor(
        source: Parcel,
        loader: ClassLoader?
    ) : this(source.readParcelable(loader)!!, source.readParcelable(loader)!!)

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeParcelable(path, flags)
        dest.writeParcelable(localFileStore, flags)
    }

    companion object {
        @JvmField
        val CREATOR = object : Parcelable.ClassLoaderCreator<LinuxFileStore> {
            override fun createFromParcel(source: Parcel): LinuxFileStore =
                createFromParcel(source, null)

            override fun createFromParcel(source: Parcel, loader: ClassLoader?): LinuxFileStore =
                LinuxFileStore(source, loader)

            override fun newArray(size: Int): Array<LinuxFileStore?> = arrayOfNulls(size)
        }
    }
}
